### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by dmtrKovalenko**
1. Move to new lifecycle methods
2. Migrate from inheritance to HOC with Render Prop #170 
3. Use new React.forwardRef to accept component by ref.

All of this issues are related to each other so I done them together. 
P.S. Maybe a little overhead with recompose, but it looks so nice

### CODE REVIEW

File: lib/src/DatePicker/Calendar.jsx [hunk 1]
```diff
@@ -62,10 +62,12 @@ export class Calendar extends Component {
     }
   }
 
-  componentWillReceiveProps(nextProps) {
-    this.setState({
-      currentMonth: this.props.utils.getStartOfMonth(nextProps.date),
-    });
+  componentDidUpdate(prevProps) {
```
[the remark this answers is not in the document]
Comment (dmtrKovalenko) [reply lib/src/DatePicker/Calendar.jsx]: @cherniavskii as far I understand - no. Here we need to determine if actually passed props changed. But there is no way to access current props in `getDerivedStateFromProps` method.

[the remark this answers is not in the document]
Comment (dmtrKovalenko) [reply lib/src/DatePicker/Calendar.jsx]: @cherniavskii but if you know better way how determine that passed props changed let me a shout)\

[the remark this answers is not in the document]
Comment (dmtrKovalenko) [reply lib/src/DatePicker/Calendar.jsx]: @cherniavskii Have you any additional considerations? I want to merge it asap)

[the remark this answers is not in the document]
Comment (cherniavskii) [reply lib/src/DatePicker/Calendar.jsx]: @dmtrKovalenko sorry for a late response, I'm pretty busy this week :/

The only way to compare new and current props in `getDerivedStateFromProps` is to store particular prop in state. But AFAIK it's recommended way to update state based on props - see example:
https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html#updating-state-based-on-props

**Comment by dmtrKovalenko** [comment]
@cherniavskii I have fixed your points. Can you please check it out one more time? Thank you

**Pull request merged**